MDEV-16330 Allow instant change of WITH SYSTEM VERSIONING column attribute

kevgs opened the pull request:
Changing columns WITH/WITHOUT SYSTEM VERSIONING doens't require to read data at
all. Thus it should be an instant operation.

Patch also fixes a bug when ALTER_COLUMN_UNVERSIONED wasn't passed to InnoDB
to change its internal structures.

innobase_change_fields_versioning(): apply WITH/WITHOUT SYSTEM VERSIONING to
table fields on ALTER TABLE.

I'm contributing this new code of the whole pull request, including one or several files that are either new files or modified ones, under the BSD-new license.

dr-m commented on storage/innobase/handler/handler0alter.cc:
```diff
@@ -8433,6 +8433,63 @@ innobase_update_foreign_cache(
+		Create_field* create_field = NULL;
+		List_iterator_fast<Create_field> it(
+		    ha_alter_info->alter_info->create_list);
+		while (Create_field* cf = it++) {
```
Could we create the `it` only once outside the `for` loop, and `it.rewind()` here inside the loop?

dr-m commented on storage/innobase/handler/handler0alter.cc:
```diff
@@ -8811,6 +8868,11 @@ commit_try_norebuild(
+		innobase_change_fields_versioning(ha_alter_info, ctx,
+						  old_table);
+	}
+
```
After this point, `commit_try_norebuild()` can still fail, and the operation may be rolled back.
I think that the data dictionary cache adjustment should be done in `commit_cache_norebuild()`. Before commit, in `commit_try_norebuild()`, we must update the `SYS_COLUMNS` only, similar to how `innobase_rename_columns_try()` and `innobase_enlarge_columns_try()` work.

dr-m commented on mysql-test/suite/versioning/t/trx_id.test:
```diff
@@ -411,3 +411,54 @@ DROP TABLE tts;
+
+--echo without table rebuild
+ALTER TABLE t
+  CHANGE a a INT WITHOUT SYSTEM VERSIONING;
```
Add `--enable_info` so that if the operation is executed using `ALGORITM=INPLACE` (as we expect), then 0 rows affected will be reported, instead of the actual number of rows that `ALGORITHM=COPY` would report.

kevgs commented:
Status update is here.

Fixed iterator (your suggestion could work faster), added tests for persistence, now changing `SYS_COLUMNS` too and found frontend bug [https://jira.mariadb.org/browse/MDEV-16490](MDEV-16490).

And right now I'm working with this test case:
```sql
set @@system_versioning_alter_history=keep;
create or replace table t (
  a int,
  b int,
  row_start bigint unsigned as row start invisible,
  row_end bigint unsigned as row end invisible,
  period for system_time(row_start, row_end)
) with system versioning engine=innodb;


insert into t values (1, 1);

select c.prtype from information_schema.innodb_sys_columns as c
  join information_schema.innodb_sys_tables as t on c.table_id=t.table_id
  where t.name='test/t' and c.name='b';
SELECT variable_value FROM INFORMATION_SCHEMA.GLOBAL_STATUS WHERE VARIABLE_NAME='innodb_instant_alter_column';

# These are from include/shutdown_mysqld.inc and allow to call start_mysqld.inc
--let $_server_id= `SELECT @@server_id`
--let $_expect_file_name= $MYSQLTEST_VARDIR/tmp/mysqld.$_server_id.expect

let $MYSQLD_DATADIR= `select @@datadir`;
let datadir= `select @@datadir`;

SET DEBUG_DBUG='+d,innodb_alter_commit_crash_before_commit';
alter table t
  change b b int without system versioning;

--source include/start_mysqld.inc
SELECT variable_value FROM INFORMATION_SCHEMA.GLOBAL_STATUS WHERE VARIABLE_NAME='innodb_instant_alter_column';
select c.prtype from information_schema.innodb_sys_columns as c
  join information_schema.innodb_sys_tables as t on c.table_id=t.table_id
  where t.name='test/t' and c.name='b';

show create table t;

select count(*) from t for system_time all;
update t set b=11;
select count(*) from t for system_time all;

set @@system_versioning_alter_history=error;
```

Here I can see an inconsistency: `SHOW CREATE TABLE` shows updated table bug `SYS_COLUMNS` is not updated and table act as if `b` is versioned (inserts historical record on `UPDATE`).

kevgs commented:
@dr-m I've updated patch and asking for another review round.

`innodb_instant_alter_column` is not affected as with column renaming and changing default values. Is it ok?

dr-m commented on mysql-test/suite/versioning/t/trx_id.test:
```diff
@@ -411,3 +411,154 @@ DROP TABLE tts;
+--disable_info
+
+--source include/shutdown_mysqld.inc
+--source include/start_mysqld.inc
```
I would prefer `restart_mysqld.inc`.

dr-m commented on mysql-test/suite/versioning/t/trx_id.test:
```diff
@@ -411,3 +411,154 @@ DROP TABLE tts;
+--disable_info
+
+--source include/shutdown_mysqld.inc
+--source include/start_mysqld.inc
```
Restarting the server is slow. Can we test everything with a single restart? (Create and alter multiple tables before the restart.)

Does the whole test need to be disabled for the embedded server? (Embedded server tests do not support restarting, but before these changes, this test did not appear to shutdown or restart the server.)

dr-m commented on mysql-test/suite/versioning/t/trx_id.test:
```diff
@@ -411,3 +411,154 @@ DROP TABLE tts;
+  JOIN INFORMATION_SCHEMA.INNODB_SYS_TABLES AS t ON C.TABLE_ID=t.TABLE_ID
+  WHERE t.NAME='test/t' AND C.NAME='b';
+
+SET DEBUG_DBUG='+d,ib_commit_inplace_fail_1';
```
Before this, I would do
```sql
SET @saved_debug_dbug = @@SESSION.debug_dbug;
```
and after the instrumented statement, restore the setting:
```sql
SET DEBUG_DBUG = @saved_debug_dbug;
```
In this way, if further tests are added to the same file, there will be no surprises due to this fault injection still being active.

dr-m commented on sql/sql_table.cc:
```diff
@@ -6537,8 +6537,7 @@ static bool fill_alter_inplace_info(THD *thd,
                                     ALTER_RENAME_COLUMN |
-                                    ALTER_CHANGE_COLUMN |
-                                    ALTER_COLUMN_UNVERSIONED));
+                                    ALTER_CHANGE_COLUMN));
```
Did you test this change with other storage engines than InnoDB?

kevgs replied:
I'll better make this for InnoDB only. Thanks for noticing.

dr-m commented on storage/innobase/handler/handler0alter.cc:
```diff
@@ -8433,6 +8433,178 @@ innobase_update_foreign_cache(
+				     "END;\n",
+				     false, trx);
+
+	trx->error_state = DB_SUCCESS;
```
Why do we blindly reset `error_state` here?

kevgs replied:
Because other parts of code does the same, f.ex. [here](https://github.com/MariaDB/server/blob/b534a7b89ef39cd75bfe09524be044bc16ea2f48/storage/innobase/handler/handler0alter.cc#L3760) and [here](https://github.com/MariaDB/server/blob/b534a7b89ef39cd75bfe09524be044bc16ea2f48/storage/innobase/handler/handler0alter.cc#L7989)

I've decided it's a preferred way to handle errors in this code. But now I can see a bit different code examples like [this one](https://github.com/MariaDB/server/blob/b534a7b89ef39cd75bfe09524be044bc16ea2f48/storage/innobase/handler/handler0alter.cc#L7687)

I'll try to understand that a bit better.

kevgs replied:
`que_eval_sql` sets `trx->error_state` and returns it. So I think it's ok to reset it and return saved value.

dr-m commented on storage/innobase/handler/handler0alter.cc:
```diff
@@ -8433,6 +8433,178 @@ innobase_update_foreign_cache(
+				     "UPDATE SYS_COLUMNS SET PRTYPE=:prtype\n"
+				     "WHERE TABLE_ID=:tableid AND NAME=:name;\n"
+				     "END;\n",
+				     false, trx);
```
I would use `POS` instead of `NAME` in the `WHERE` condition, because the primary key of `SYS_COLUMNS` is `(TABLE_ID,POS)`.

Ideally we should check that exactly one record was updated.

dr-m commented on storage/innobase/handler/handler0alter.cc:
```diff
@@ -8433,6 +8433,178 @@ innobase_update_foreign_cache(
+	trx->error_state = DB_SUCCESS;
+	trx->op_info = "";
+
+	DBUG_RETURN(error != DB_SUCCESS);
```
If `ha_innobase::commit_inplace_alter_table()` returns `true`, it must have invoked `my_error()`. Code like this is present elsewhere:
```c++
	if (error != DB_SUCCESS) {
		my_error_innodb(error, table_name, 0);
		trx->error_state = DB_SUCCESS;
		trx->op_info = "";
		DBUG_RETURN(true);
	}
```

dr-m commented on storage/innobase/handler/handler0alter.cc:
```diff
@@ -8433,6 +8433,178 @@ innobase_update_foreign_cache(
+		if (create_field->versioning
+		    == Column_definition::WITHOUT_VERSIONING) {
+			DBUG_ASSERT(!col->vers_sys_start()
+				    && !col->vers_sys_end());
```
Never use `&&` at the top level in assertion expressions. Instead, split the assertion:
```c++
DBUG_ASSERT(!col->vers_sys_start());
DBUG_ASSERT(!col->vers_sys_end());
```

dr-m commented on storage/innobase/handler/handler0alter.cc:
```diff
@@ -8433,6 +8433,178 @@ innobase_update_foreign_cache(
+@retval false	on success */
+static
+bool
+innobase_change_field_versioning_try(
```
I would omit the `innobase_` prefix from the names of the added functions.

dr-m commented on mysql-test/suite/versioning/t/trx_id.test:
```diff
@@ -411,3 +411,154 @@ DROP TABLE tts;
+
+INSERT INTO t VALUES (1,1);
+
+--echo without table rebuild
```
Start this kind of `--echo` output with `#`, so that the comment will better be distinguished from SQL statements and their output in the .result file.

dr-m commented on mysql-test/suite/versioning/t/trx_id.test:
```diff
@@ -411,3 +411,154 @@ DROP TABLE tts;
+--enable_info
+ALTER TABLE t
+  CHANGE a a INT WITH SYSTEM VERSIONING,
+  ADD PRIMARY KEY pk (a);
```
Can you first try the same with `ALGORITHM=INSTANT`? That variant should be rejected for a rebuilding operation, but should be accepted in the simple case.

kevgs commented:
> Ideally we should check that exactly one record was updated.

This was not done. But I found that `MONITOR_OLVD_SYSTEM_ROW_UPDATED` is never updated. And `srv_stats.n_system_rows_updated` is not updated when f.ex. `SYS_COLUMNS` updated by `que_run_sql()` because it updated only in `row_update_for_mysql()`. Is these a bugs?

The rest is fixed.

dr-m commented on mysql-test/suite/versioning/t/trx_id_versioning_attribute_persistence.test:
```diff
@@ -0,0 +1,86 @@
+--enable_info
+--echo # without rebuild
+ALTER TABLE t1
+  CHANGE a a INT WITHOUT SYSTEM VERSIONING;
```
Add `ALGORITHM=INSTANT` also here.

dr-m commented on storage/innobase/handler/handler0alter.cc:
```diff
@@ -8433,6 +8433,188 @@ innobase_update_foreign_cache(
+change_field_versioning_try(
+	trx_t* trx,
+	const char* table_name,
+	const ulint tableid,
```
`tableid` should be `table_id_t` (64 bits). `ulint` is only 32 bits on 32-bit systems.

kevgs replied:
That also caused a bb failure for windows.

dr-m commented on storage/innobase/handler/handler0alter.cc:
```diff
@@ -8433,6 +8433,188 @@ innobase_update_foreign_cache(
+
+		if (change_field_versioning_try(
+			trx, table->s->table_name.str,
+			static_cast<ulint>(new_table->id), pos, new_prtype)) {
```
Remove the cast. We must pass all 64 bits of `dict_table_t::id` also on 32-bit systems.

The formula of `pos` looks OK to me. For `SYS_FIELDS.POS` it would encode a prefix length, but `SYS_COLUMNS.POS` is a simple number for non-virtual columns.

dr-m commented:
>This was not done. But I found that MONITOR_OLVD_SYSTEM_ROW_UPDATED is never updated.

It looks like this counter as well as `srv_stats.n_system_rows_updated` were introduced in commit caeffc7a7d2b259c33a4512caa951fddb369a8b5. Indeed, I do not see a test where the `system` counters would change. In my opinion, either the non-working change should be reverted, or it should be fixed.

>And `srv_stats.n_system_rows_updated` is not updated when f.ex. `SYS_COLUMNS` updated by `que_eval_sql()` because it updated only in `row_update_for_mysql()`. Is these a bugs?

That is not a bug in my opinion. According to the commit comment, the InnoDB internal tables do not count as system tables. Furthermore, I would like to remove the InnoDB system tables.

dr-m commented:
Pushed to 10.3 as commit 9dc81f7d387050dd62f2307b15c63c3a3f5ea1b0

kevgs commented:
@dr-m 

Here is a list of `srv_stat_t` variables and example tests where it's used:
`n_system_rows_read` - `innodb.innodb-index-online`
`n_system_rows_updated` - `innodb.innodb_stats_fetch`
`n_system_rows_deleted` - `innodb.innodb_stats_fetch_nonexistent`
`n_system_rows_inserted` - `innodb.innodb-index-online`

And these are not used anywhere:
```c++
	MONITOR_OLVD_SYSTEM_ROW_READ,
	MONITOR_OLVD_SYSTEM_ROW_INSERTED,
	MONITOR_OLVD_SYSTEM_ROW_DELETED,
	MONITOR_OLVD_SYSTEM_ROW_UPDATED,
```

dr-m commented:
@kevgs thank you! Can you file MDEV and a pull request for removing the untested instrumentation? If the `MONITOR` instrumentation for these does not work at all, then we could remove it already in GA versions; else we can only remove it in 10.4.
